solutions/gold/baltic-09-candy.mdx

```diff
@@ -0,0 +1,156 @@
+wagon will be at $s_j$ to catch $j$). However, at time $t_j$ the range for
+acceptable wagon positions of $c_i$ will have been reduced to
+$$
+[s_i - (t_i - t_j), s_i + (t_i - t_j)]
+$$
+
+But $s_j$ needs to be inside this range, therefore:
+$$
+s_i - (t_i - t_j) \leq s_j \text{ and } s_j \leq s_i + (t_i - t_j) \iff
+$$
+$$
+s_i - t_i \leq s_j - t_j \text{ and } s_j + t_j \leq s_i + t_i
+$$
+
+This is equivalent to saying that the range of acceptable wagon positions of
+$c_j$ is fully contained within the range of $c_i$.
+
+We have therefore concluded that two candies can be caught by the same wagon
+if (and only if) their ranges of acceptable wagon positions are nested.
+
+### Finding the minimum number of wagons
+
+We can think of each candy as its range of acceptable wagon positions at time
+0 $[s - t, s + t]$ (note that each candy corresponds to a unique range and vice
+versa). We can conclude that a wagon can catch a sequence of candies
+$c_1, c_2, ..., c_k$, if they form a sequence of nested ranges. It follows that
+the minimum number of wagons needed is the longest sequence of non-nested
+ranges.
+
+Two ranges $[l_1, r_1]$, $[l_2, r_2]$ that are not nested adhere to:
+$$l_1 < l_2 \implies r_1 < r_2$$.
+
+If we sort the candies by the left endpoint of their ranges such that
+$$
+l_1 \leq l_2 \leq ... \leq l_n
+$$
+then a sequence of non-nested ranges would correspond to an increasing
+sequence of right endpoints. Therefore, the minimum number of required wagons
+would correspond to the length of the longest increasing sequence of right
+endpoints.
+
+### Assigning the candies to wagons
+
+By following the binary search algorithm described in
+[the LIS module](/gold/lis), we can assign the candies with the same LIS length
+to the same wagon. The ranges that have the same LIS length cannot be nested.
+If they were, we could use the inside range to increase the LIS of the outside
+one by 1.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N \log N)$
+
+<Warning>
+Watch out for the case $l_i = l_j$, in which case we must sort by decreasing
+right endpoint to ensure they would not be counted as non-nested ranges.
+</Warning>
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+typedef pair<int, int> pi;
+
+int main() {
+	int n;
+	cin >> n;
+
+	vector<pi> candies(n);
+	vector<int> lis(n + 1);
+	vector<vector<pi>> candies_caught(n + 1);
+
+	for (int i = 0; i < n; i++) {
+		int s, t;
+		cin >> s >> t;
+		candies[i] = make_pair(s - t, s + t);
+	}
+
+	/*
+	 * Sort the segments by left endpoint
+	 * NOTE: It is important to sort segments of same left point by decreasing
+	 * right point so they count as nested
+	 */
+	sort(candies.begin(), candies.end(), [](const pi &a, const pi &b) -> bool {
+		return (a.first == b.first ? a.second > b.second : a.first < b.first);
+	});
+
+	for (int len = 0; len <= n; len++) { lis[len] = INT_MAX; }
```

Comment:
delete this & move to `vector<int> lis(n + 1, INT_MAX);`?

completely optional though